library/bigip_asm_policy.py

```diff
@@ -0,0 +1,546 @@
+                    map_key = k
+
+                # Handle weird API parameters like `dns.proxy.__iter__` by
+                # using a map provided by the module developer
+                class_attr = getattr(type(self), map_key, None)
+                if isinstance(class_attr, property):
+                    # There is a mapped value for the api_map key
+                    if class_attr.fset is None:
+                        # If the mapped value does not have
+                        # an associated setter
+                        self._values[map_key] = v
+                    else:
+                        # The mapped value has a setter
+                        setattr(self, map_key, v)
+                else:
+                    # If the mapped value is not a @property
+                    self._values[map_key] = v
+
+    updatables = [
+        'active'
+    ]
+
+    returnables = [
+        'name', 'template', 'file', 'active'
+    ]
+
+    api_attributes = [
+        'name', 'file'
+    ]
+    api_map = {
+        'filename': 'file'
+    }
+
+    @property
+    def template(self):
+        tmpl_name = self._values['template']
+        if tmpl_name is None:
+            return None
+        if self._template_exists_on_device(tmpl_name):
+            return self._values['template']
+        else:
+            raise F5ModuleError('Template with the given name: {0} does not exist'.format(tmpl_name))
+
+    def _template_exists_on_device(self, name):
+        collection = self._templates_on_device()
+        for resource in collection:
+            if resource.name == name.upper():
+                self._set_template_selflink(resource)
+                return True
+        return False
+
+    def _set_template_selflink(self, template):
+        link = {'link': template.selfLink}
+        self._values['template_link'] = link
+
+    def _templates_on_device(self):
+        collection = self.client.api.tm.asm.policy_templates_s.get_collection()
+        return collection
+
+    def to_return(self):
+        result = {}
+        for returnable in self.returnables:
+            result[returnable] = getattr(self, returnable)
+        result = self._filter_params(result)
+        return result
+
+    def api_params(self):
+        result = {}
+        for api_attribute in self.api_attributes:
+            if self.api_map is not None and api_attribute in self.api_map:
+                result[api_attribute] = getattr(self,
+                                                self.api_map[api_attribute])
+            else:
+                result[api_attribute] = getattr(self, api_attribute)
+        result = self._filter_params(result)
+        return result
+
+
+class ModuleManager(object):
+    def __init__(self, client):
+        self.client = client
+        self.have = None
+        self.want = Parameters()
+        self.want.client = self.client
+        self.want.update(self.client.module.params)
+        self.changes = Parameters()
```

Comment:
change this to `Changes()` as we use a changes class to hold these changes. This additionally helps make callgraphs easier to read